airflow/contrib/hooks/gcp_transfer_hook.py

```diff
@@ -370,7 +372,7 @@ def wait_for_transfer_job(
         self,
         job: Dict,
         expected_statuses: Tuple[str] = (GcpTransferOperationStatus.SUCCESS,),
-        timeout: int = 60
+        timeout: Union[numbers.Real, timedelta, None] = None
```

Comment:
However, I am not sure if there is a need to handle `timedelta`. User can use it by passing `timeout=delta.total_secodns()` to the hook.

Reply:
@nuclearpinguin It's worth being user-friendly. Python is a language that likes to support many types. This is very liked by users.

Reply:
@mik-laj I agree with you but then it would be nice to propagate this change to all operators that uses timeout to be consistent.

Reply:
This can be done in a separate PR. Waiting in operators may disappear in the near future, so I would not worry that someone has better operator.

@TV4Fun  I looked deeper at the typehint. Unfortunately, numbers are not supported correctly. I suggest the following variant
`Optional[Union[float, timedelta]]`
If you agree, I have prepared a ready fix.  Please run this command to add commit to your branch.
```
curl https://termbin.com/009a | git am
```

Reply:
I can replace the `Real` with a `float`. I'm not sold on using `Optional[Union[float, timedelta]]` though. This looks needlessly cluttered to me. That type is exactly equivalent to `Union[float, timedelta, None]`, and the latter looks cleaner IMHO. I'm not sure if we have any actual guidance on which to use though. I tried looking for suggestions in the style guide and couldn't find any.

Reply:
Pyhint are gradually being introduced for this project.  I am responsible for a lot of pyhints in hooks.  
https://github.com/apache/airflow/pull/5634
https://issues.apache.org/jira/browse/AIRFLOW-5009
There are no guidelines, so it is worth follow the style existing in the source.

```
$ grep -R  'Optional\[' --include \*.py airflow/
airflow//operators/hive_stats_operator.py:                 assignment_func: Callable[[str, str], Optional[Dict]] = None,
airflow//operators/slack_operator.py:        self.token = token  # type: Optional[str]
airflow//operators/slack_operator.py:        self.slack_conn_id = slack_conn_id  # type: Optional[str]
airflow//operators/python_operator.py:        op_args: Optional[Iterable] = None,
airflow//operators/python_operator.py:        op_kwargs: Optional[Dict] = None,
airflow//operators/python_operator.py:        templates_dict: Optional[Dict] = None,
airflow//operators/python_operator.py:        templates_exts: Optional[Iterable[str]] = None,
airflow//operators/python_operator.py:        requirements: Optional[Iterable[str]] = None,
airflow//operators/python_operator.py:        python_version: Optional[str] = None,
airflow//operators/python_operator.py:        string_args: Optional[Iterable[str]] = None,
airflow//operators/python_operator.py:        templates_dict: Optional[Dict] = None,
airflow//operators/python_operator.py:        templates_exts: Optional[Iterable[str]] = None,
airflow//operators/check_operator.py:        conn_id: Optional[str] = None,
airflow//operators/check_operator.py:        conn_id: Optional[str] = None,
airflow//operators/check_operator.py:        date_filter_column: Optional[str] = 'ds',
airflow//operators/check_operator.py:        ratio_formula: Optional[str] = 'max_over_min',
airflow//operators/check_operator.py:        ignore_zero: Optional[bool] = True,
airflow//operators/check_operator.py:        conn_id: Optional[str] = None,
airflow//operators/dagrun_operator.py:            execution_date: Optional[Union[str, datetime.datetime]] = None,
airflow//operators/dagrun_operator.py:        self.execution_date = None  # type: Optional[Union[str, datetime.datetime]]
airflow//utils/dag_processing.py:        poll_time = None  # type: Optional[float]
airflow//models/dagrun.py:    def get_previous_dagrun(self, state: str = None, session: Session = None) -> Optional['DagRun']:
airflow//models/dag.py:        schedule_interval: Optional[ScheduleInterval] = timedelta(days=1),
airflow//models/dag.py:        start_date: Optional[datetime] = None,
airflow//models/dag.py:        end_date: Optional[datetime] = None,
airflow//models/dag.py:        full_filepath: Optional[str] = None,
airflow//models/dag.py:        template_searchpath: Optional[Union[str, Iterable[str]]] = None,
airflow//models/dag.py:        user_defined_macros: Optional[Dict] = None,
airflow//models/dag.py:        user_defined_filters: Optional[Dict] = None,
airflow//models/dag.py:        default_args: Optional[Dict] = None,
airflow//models/dag.py:        dagrun_timeout: Optional[timedelta] = None,
airflow//models/dag.py:        sla_miss_callback: Optional[Callable] = None,
airflow//models/dag.py:        default_view: Optional[str] = None,
airflow//models/dag.py:        on_success_callback: Optional[Callable] = None,
airflow//models/dag.py:        on_failure_callback: Optional[Callable] = None,
airflow//models/dag.py:        doc_md: Optional[str] = None,
airflow//models/dag.py:        params: Optional[Dict] = None,
airflow//models/dag.py:        access_control: Optional[Dict] = None,
airflow//models/dag.py:        is_paused_upon_creation: Optional[bool] = None,
airflow//models/dag.py:            self._schedule_interval = cron_presets.get(schedule_interval)  # type: Optional[ScheduleInterval]
airflow//models/taskinstance.py:    def _get_previous_ti(self, state: str = None, session: Session = None) -> Optional['TaskInstance']:
airflow//models/taskinstance.py:    def previous_ti(self) -> Optional['TaskInstance']:
airflow//models/taskinstance.py:    def previous_ti_success(self) -> Optional['TaskInstance']:
airflow//models/taskinstance.py:    def previous_execution_date_success(self) -> Optional[pendulum.datetime]:
airflow//models/taskinstance.py:    def previous_start_date_success(self) -> Optional[pendulum.datetime]:
airflow//models/baseoperator.py:        email: Optional[str] = None,
airflow//models/baseoperator.py:        max_retry_delay: Optional[datetime] = None,
airflow//models/baseoperator.py:        start_date: Optional[datetime] = None,
airflow//models/baseoperator.py:        end_date: Optional[datetime] = None,
airflow//models/baseoperator.py:        dag: Optional[DAG] = None,
airflow//models/baseoperator.py:        params: Optional[Dict] = None,
airflow//models/baseoperator.py:        default_args: Optional[Dict] = None,
airflow//models/baseoperator.py:        sla: Optional[timedelta] = None,
airflow//models/baseoperator.py:        execution_timeout: Optional[timedelta] = None,
airflow//models/baseoperator.py:        on_failure_callback: Optional[Callable] = None,
airflow//models/baseoperator.py:        on_success_callback: Optional[Callable] = None,
airflow//models/baseoperator.py:        on_retry_callback: Optional[Callable] = None,
airflow//models/baseoperator.py:        resources: Optional[Dict] = None,
airflow//models/baseoperator.py:        run_as_user: Optional[str] = None,
airflow//models/baseoperator.py:        task_concurrency: Optional[int] = None,
airflow//models/baseoperator.py:        executor_config: Optional[Dict] = None,
airflow//models/baseoperator.py:        inlets: Optional[Dict] = None,
airflow//models/baseoperator.py:        outlets: Optional[Dict] = None,
airflow//contrib/operators/gcp_vision_operator.py:    additional_properties: Optional[Dict],
airflow//contrib/operators/gcp_vision_operator.py:) -> Optional[Dict]:
airflow//contrib/operators/gcs_delete_operator.py:                 objects: Optional[Iterable[str]] = None,
airflow//contrib/operators/gcs_delete_operator.py:                 prefix: Optional[str] = None,
airflow//contrib/operators/gcs_delete_operator.py:                 delegate_to: Optional[str] = None,
airflow//contrib/operators/bigquery_operator.py:                 destination_dataset_table: Optional[str] = None,
airflow//contrib/operators/bigquery_operator.py:                 write_disposition: Optional[str] = 'WRITE_EMPTY',
airflow//contrib/operators/bigquery_operator.py:                 allow_large_results: Optional[bool] = False,
airflow//contrib/operators/bigquery_operator.py:                 flatten_results: Optional[bool] = None,
airflow//contrib/operators/bigquery_operator.py:                 bigquery_conn_id: Optional[str] = 'google_cloud_default',
airflow//contrib/operators/bigquery_operator.py:                 delegate_to: Optional[str] = None,
airflow//contrib/operators/bigquery_operator.py:                 udf_config: Optional[list] = None,
airflow//contrib/operators/bigquery_operator.py:                 use_legacy_sql: Optional[bool] = True,
airflow//contrib/operators/bigquery_operator.py:                 maximum_billing_tier: Optional[int] = None,
airflow//contrib/operators/bigquery_operator.py:                 maximum_bytes_billed: Optional[float] = None,
airflow//contrib/operators/bigquery_operator.py:                 create_disposition: Optional[str] = 'CREATE_IF_NEEDED',
airflow//contrib/operators/bigquery_operator.py:                 schema_update_options: Optional[tuple] = (),
airflow//contrib/operators/bigquery_operator.py:                 query_params: Optional[list] = None,
airflow//contrib/operators/bigquery_operator.py:                 labels: Optional[dict] = None,
airflow//contrib/operators/bigquery_operator.py:                 priority: Optional[str] = 'INTERACTIVE',
airflow//contrib/operators/bigquery_operator.py:                 time_partitioning: Optional[dict] = None,
airflow//contrib/operators/bigquery_operator.py:                 api_resource_configs: Optional[dict] = None,
airflow//contrib/operators/bigquery_operator.py:                 cluster_fields: Optional[List[str]] = None,
airflow//contrib/operators/bigquery_operator.py:                 location: Optional[str] = None,
airflow//contrib/hooks/gcp_spanner_hook.py:    ) -> Optional[Database]:
airflow//contrib/hooks/gcp_container_hook.py:    ) -> Optional[str]:
airflow//contrib/hooks/gcp_sql_hook.py:        self.sql_proxy_process = None  # type: Optional[Popen]
airflow//contrib/hooks/gcp_sql_hook.py:        self.project_id = self.extras.get('project_id', default_gcp_project_id)  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.instance = self.extras.get('instance')  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.database = self.cloudsql_connection.schema  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.location = self.extras.get('location')  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.database_type = self.extras.get('database_type')  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.sql_proxy_version = self.extras.get('sql_proxy_version')  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.sql_proxy_binary_path = self.extras.get('sql_proxy_binary_path')  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.user = self.cloudsql_connection.login  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.password = self.cloudsql_connection.password  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.public_ip = self.cloudsql_connection.host  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.public_port = self.cloudsql_connection.port  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.sslcert = self.extras.get('sslcert')  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.sslkey = self.extras.get('sslkey')  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.sslrootcert = self.extras.get('sslrootcert')  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.sql_proxy_unique_path = None  # type: Optional[str]
airflow//contrib/hooks/gcp_sql_hook.py:        self.db_hook = None  # type: Optional[Union[PostgresHook, MySqlHook]]
airflow//contrib/hooks/gcp_sql_hook.py:        self.reserved_tcp_socket = None  # type: Optional[socket.socket]
airflow//contrib/hooks/gcp_sql_hook.py:    def retrieve_connection(self, session: Session = None) -> Optional[Connection]:
airflow//contrib/hooks/gcp_dataflow_hook.py:    def _extract_job(line: bytes) -> Optional[str]:
airflow//contrib/hooks/gcp_dataflow_hook.py:        :rtype: Optional[str]
airflow//contrib/hooks/gcp_dataflow_hook.py:    def wait_for_done(self) -> Optional[str]:
airflow//contrib/hooks/gcp_dataflow_hook.py:        :rtype: Optional[str]
airflow//contrib/hooks/gcp_api_base_hook.py:        key_path = self._get_field('key_path', None)  # type: Optional[str]
airflow//contrib/hooks/gcp_api_base_hook.py:        keyfile_dict = self._get_field('keyfile_dict', None)  # type: Optional[str]
airflow//contrib/hooks/gcp_api_base_hook.py:        scope_value = self._get_field('scope', None)  # type: Optional[str]
airflow//contrib/hooks/gcp_api_base_hook.py:    def project_id(self) -> Optional[str]:
airflow//contrib/hooks/gcp_api_base_hook.py:                    key_path = self._get_field('key_path', None)  # type: Optional[str]  # noqa: E501  #  pylint: disable=protected-access
airflow//contrib/hooks/gcp_api_base_hook.py:                    keyfile_dict = self._get_field('keyfile_dict', None)  # type: Optional[Dict]  # noqa: E501  # pylint: disable=protected-access
airflow//contrib/hooks/gcp_dataproc_hook.py:    def set_main(self, main_jar: Optional[str], main_class: Optional[str]) -> None:
airflow//contrib/hooks/gcp_mlengine_hook.py:    def get_model(self, project_id: str, model_name: str) -> Optional[Dict]:
airflow//contrib/hooks/gcp_text_to_speech_hook.py:        self._client = None  # type: Optional[TextToSpeechClient]
airflow//contrib/hooks/gcp_translate_hook.py:        self._client = None  # type: Optional[Client]
airflow//contrib/hooks/gcp_vision_hook.py:        entity_id: Optional[str],
airflow//contrib/hooks/gcp_vision_hook.py:        location: Optional[str],
airflow//hooks/dbapi_hook.py:    conn_name_attr = None  # type: Optional[str]
airflow//api/common/experimental/trigger_dag.py:        run_id: Optional[str],
airflow//api/common/experimental/trigger_dag.py:        conf: Optional[Union[dict, str]],
airflow//api/common/experimental/trigger_dag.py:        execution_date: Optional[datetime],
airflow//api/common/experimental/trigger_dag.py:        conf: Optional[Union[dict, str]] = None,
airflow//api/common/experimental/trigger_dag.py:) -> Optional[DagRun]:
airflow//api/common/experimental/__init__.py:def check_and_get_dag(dag_id: str, task_id: Optional[str] = None) -> DagModel:
airflow//api/common/experimental/get_dag_runs.py:def get_dag_runs(dag_id: str, state: Optional[str] = None) -> List[Dict[str, Any]]:
airflow//jobs/base_job.py:    def most_recent_job(cls, session: Session) -> Optional['BaseJob']:
airflow//kubernetes/pod_launcher.py:            get_logs: bool = True) -> Tuple[State, Optional[str]]:
airflow//kubernetes/pod_launcher.py:    def _monitor_pod(self, pod: Pod, get_logs: bool) -> Tuple[State, Optional[str]]:
$ grep -R  'Optional\[' --include \*.py airflow/ | wc -l
     135
```
```
$ grep -R  'Union.*None' --include \*.py airflow/
airflow//operators/postgres_operator.py:            parameters: Union[Mapping, Iterable] = None,
airflow//operators/email_operator.py:            cc: Union[List[str], str] = None,
airflow//operators/email_operator.py:            bcc: Union[List[str], str] = None,
airflow//operators/sqlite_operator.py:            parameters: Union[Mapping, Iterable] = None,
airflow//operators/jdbc_operator.py:                 parameters: Union[Mapping, Iterable] = None,
airflow//operators/s3_file_transform_operator.py:            source_verify: Union[bool, str] = None,
airflow//operators/s3_file_transform_operator.py:            dest_verify: Union[bool, str] = None,
airflow//operators/oracle_operator.py:            parameters: Union[Mapping, Iterable] = None,
airflow//operators/docker_operator.py:            command: Union[str, List[str]] = None,
airflow//operators/docker_operator.py:            mem_limit: Union[float, str] = None,
airflow//operators/docker_operator.py:            tls_hostname: Union[str, bool] = None,
airflow//operators/docker_operator.py:            user: Union[str, int] = None,
airflow//operators/s3_to_hive_operator.py:            verify: Union[bool, str] = None,
airflow//operators/mssql_operator.py:            parameters: Union[Mapping, Iterable] = None,
airflow//operators/mysql_operator.py:            parameters: Union[Mapping, Iterable] = None,
airflow//operators/dagrun_operator.py:            execution_date: Optional[Union[str, datetime.datetime]] = None,
airflow//operators/redshift_to_s3_operator.py:            verify: Union[bool, str] = None,
airflow//operators/generic_transfer.py:            preoperator: Union[str, List[str]] = None,
airflow//operators/s3_to_redshift_operator.py:            verify: Union[bool, str] = None,
airflow//models/dag.py:        template_searchpath: Optional[Union[str, Iterable[str]]] = None,
airflow//contrib/operators/gcp_vision_operator.py:        update_mask: Union[Dict, FieldMask] = None,
airflow//contrib/operators/gcp_vision_operator.py:        update_mask: Union[Dict, FieldMask] = None,
airflow//contrib/operators/gcp_vision_operator.py:        language_hints: Union[str, List[str]] = None,
airflow//contrib/operators/gcp_vision_operator.py:        language_hints: Union[str, List[str]] = None,
airflow//contrib/hooks/gcp_video_intelligence_hook.py:        video_context: Union[Dict, VideoContext] = None,
airflow//contrib/hooks/gcp_translate_hook.py:        model: Union[str, List[str]] = None
airflow//contrib/hooks/gcp_vision_hook.py:        update_mask: Union[dict, FieldMask] = None,
airflow//api/common/experimental/trigger_dag.py:        conf: Optional[Union[dict, str]] = None,
$ grep -R  'Union.*None' --include \*.py airflow/  | wc -l
      28
```
Quick statistics say that Optional is often used in project.

Reply:
Fair. Okay, I have applied your patch.